### ISSUE TITLE
cannot catch error from fetch call when used a malformed URL

### ISSUE THREAD

**Issue by Trenrod** [facebook/react-native#18087]
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

Environment:
  OS: Windows 10
  Node: 6.12.3
  Yarn: Not Found
  npm: 3.10.10
  Watchman: Not Found
  Xcode: N/A
  Android Studio: Version  3.0.0.0 AI-171.4443003

Packages: (wanted => installed)
  react: 16.2.0 => 16.2.0
  react-native: 0.53.3 => 0.53.3

Target Platform: Android 6 Emulator (From the Getting Started howto)

### Steps to Reproduce

Cannot catch the fetch call when used a malformed url.

1. Call fetch() with a malformed url like 'asd' or 'https:://facebook.github.io'
2. try to catch with '.catch' when called as promise or try/catch within async/await

### Expected Behavior

catch should be executed

### Actual Behavior

red screen shows up with "unexpected url: [URL USED WITHIN THE CALL]"

### Reproducible Demo

```
  async asyncFetch(url) {
    try {
      console.log('DEBUG - START FETCHING');
      const response = await fetch(url);
      console.log('DEBUG - END FETCHING');
      const xml = await response.text();
      const json = xml2json.parse(xml);
      return { data: json };
    } catch (error) {
      console.log('DEBUG - START ERROR: ' + error.message);
      throw error;
    }
  }

  const url = 'httasd:afdasdfads';
  this.asyncFetch(url)
  .catch(err => {
    console.log('DEBUG - Error: ' + err.message);
  })
```

Actual console output is:
DEBUG - START FETCHING

Note: Same issue like #17843

**Comment by ksegla** [comment]
fetch has been having this very bad problem for some time now, crashing apps in production whenever something it doesn't like happen. At this point, I'm looking for good alternatives :-(

**Comment by Dorentin1997** [comment]
Check if url is good and returns something. Send an request at that link with postman check what it returns.

**Comment by Trenrod** [comment]
@Dorentin1997 thank you for your comment.
Yes, a workaround can be using  'url-parse' to validate the url.
But my assumption is that if someone enters an invalid url, fetch should throw a handleble exception and not crashing the whole application.

**Comment by Dorentin1997** [comment]
Use url-parse do the logic  there,  if it`s wrong throw the exception there.

**Comment by Trenrod** [comment]
@Dorentin1997 again. Yes as a workaround.
Still this bug exists in the fetch implementation, which should be fixed.

**Comment by Dorentin1997** [comment]
@Trenrod  Yes it`s clearly a bug, i had problems too with this. Hopefully it will be fixed.

**Comment by jdnichollsc** [comment]
Please fix this issue, we're having a lot of problems with **HTTP** requests using `fetch` or `axios` (crashing apps in production too) :(

Example:
```
async login(){
  this.setState({ loading: true })
  try {
    const response = await fetch(`${configs.baseUrl}/auth/login`)
    const data = await response.text()
    //...
  } catch (error) { //The catch doesn't work, also using Promises...
    this.setState({ loading: false })
    Alert.alert(error)
  }
}
```

<img width="559" alt="captura de pantalla 2018-06-27 a la s 10 42 49 a m" src="https://user-images.githubusercontent.com/2154886/42004636-5e85ccfa-7a36-11e8-967d-cb811fbc1c02.png">

Any help is really appreciated, Juan
